cassiopeia/data/CassiopeiaTree.py

```diff
@@ -1334,14 +1568,149 @@ def filter_nodes(self, condition: Callable[[str], bool]) -> List[str]:
+        """
+        self.__check_network_initialized()
+
+        if (
+            "distances" not in self.__cache
+            or node1 not in self.__cache["distances"]
+            or node2 not in self.__cache["distances"][node1]
+        ):
+            distances = self.__cache.setdefault("distances", {}).setdefault(
+                node1, {}
+            )
+            if node1 == node2:
+                distances[node2] = 0
+            else:
+                lca = self.find_lca(node1, node2)
+                lca_time = self.get_time(lca)
+                distances[node2] = (self.get_time(node1) - lca_time) + (
+                    self.get_time(node2) - lca_time
+                )
+
+        return self.__cache["distances"][node1][node2]
+
+    def get_distances(
+        self, node: str, leaves_only: bool = False
+    ) -> Dict[str, float]:
+        """Compute the branch distance between the given node and every other
+        node. This function does NOT use :func:`get_distance` internally. Instead,
+        it uses properties of trees to compute these much quicker, and the results
+        are cached.
+
+        Args:
+            node: Node from which to compute distance to all other nodes
+            leaves_only: Calculate distances to leaves only and not internal nodes.
+                Defaults to False.
+
+        Returns:
+            Dictionary of distances
+
+        Raises:
+            CassiopeiaTreeError if the tree has not been initialized.
+        """
+        self.__check_network_initialized()
+
+        if (
+            "distances" not in self.__cache
+            or node not in self.__cache["distances"]
+            or any(
+                _node not in self.__cache["distances"][node]
+                for _node in self.nodes
+            )
+        ):
+            distances = self.__cache.setdefault("distances", {}).setdefault(
+                node, {}
+            )
+
+            # First, compute distances to descendents of this node
+            node_time = self.get_time(node)
+            for descendant in self.depth_first_traverse_nodes(node):
+                distances[descendant] = self.get_time(descendant) - node_time
+
+            # Then, compute distances to other nodes by traversing parents one by one
+            if not self.is_root(node):
+                for ancestor in self.get_all_ancestors(node):
+                    ancestor_time = self.get_time(ancestor)
+                    ancestor_distance = node_time - ancestor_time
+                    distances[ancestor] = ancestor_distance
+
+                    for descendant in self.depth_first_traverse_nodes(
+                        ancestor, postorder=False
+                    ):
+                        if descendant not in distances:
+                            distances[descendant] = ancestor_distance + (
+                                self.get_time(descendant) - ancestor_time
+                            )
```

Comment:
Isn't this code block going to reconsider nodes several times? Why not just do a single DFS from the root, ignoring descendants of the query `node` passed into the function?

Reply:
Changed this to use `nx. shortest_path_length` instead.